### PR TITLE
Workflow Option to retry jobs in JES

### PR DESCRIPTION
**Pull request by ruchim**
Taking advantage of the existing framework of "unexpectedRetryCount" and to implement https://github.com/broadinstitute/cromwell/issues/3161.

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/preparation/JobPreparationActor.scala [hunk 1]
```diff
@@ -187,7 +187,10 @@ class JobPreparationActor(workflowDescriptor: EngineWorkflowDescriptor,
                                                     runtimeAttributes: Map[LocallyQualifiedName, WomValue],
                                                     maybeCallCachingEligible: MaybeCallCachingEligible,
                                                     prefetchedJobStoreEntries: Map[String, KvResponse]): BackendJobPreparationSucceeded = {
-    val jobDescriptor = BackendJobDescriptor(workflowDescriptor.backendDescriptor, jobKey, runtimeAttributes, inputEvaluation, maybeCallCachingEligible, prefetchedJobStoreEntries)
+    val backendWorkflowDescriptor = workflowDescriptor.backendDescriptor
+    val jobDescriptor = BackendJobDescriptor(backendWorkflowDescriptor, jobKey, runtimeAttributes, inputEvaluation,
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/preparation/JobPreparationActor.scala]: ToL: IMO if you're going to separate lines, separate everything onto their own lines...

**Comment by cjllanwarne** [comment]
I'm not sure the exact scenario people would want to use this in, but it seems to replace the previous behavior "retry a few times for known flakinesses, otherwise fail" - with a less tuned "retry a custom number of times for all codes".

* Is there any possibility that users would want to continue retrying on known flakinesses (hint: I really think they do) but not blindly restart on other error codes (which could be expensive for typos!)?
---
One other thought:
* This seems to only catch JES telling us the task has failed. If the task on JES "succeeds" but Cromwell later finds out that it cannot download or evaluate a result, does this still retry?

**Comment by geoffjentry** [comment]
@cjlanwarne re discussing granularity of retries, note that this exactly satisfies a request from mint (see the ticket linked in the pr description)

**Comment by geoffjentry** [comment]
Or at least that was the intention, I haven’t looked at the code to see if it succeeds but I have faith in @ruchim :)

**Comment by geoffjentry** [comment]
Deleted a few comments as I misread what @cola Warner said. Sorry I’m leaving that autocorrect typo as it’s awesome

We should discuss what this does and does not do. I viewed the request to be an enhancement on current behavior, so should ensure that’s the case

**Comment by geoffjentry** [comment]
@cjllanwarne @ruchim I've been going back and forth as to whether or not this retry count should replace or add on to the "we know that's flakey so retried it for ya" situation vs. adding on top of it. Chris, you seemed pretty sure it should be the latter, why do you view it that way? Ruchi, I know this was spec'd out to go the former route, but I think the person speccing it out didn't know that we already did some stuff - do you actually have an opinion here or is this just how you did it?

**Comment by ruchim** [comment]
@geoffjentry @cjllanwarne I've already made changes after a conversation with Chris on Friday--so I'll close this out for now, and re-open when we've all had a chance to talk about specifics.

**Pull request closed without merging**